src/control/ros/Conversions.cpp

```diff
@@ -0,0 +1,283 @@
+  const Eigen::VectorXd& _nextVelocity,
+  const Eigen::VectorXd& _nextAcceleration,
+  size_t _numCoefficients)
+{
+  using aikido::util::SplineProblem;
+
+  assert(_numCoefficients == 2 || _numCoefficients == 4 || _numCoefficients == 6);
+
+  const auto numDofs = _currPosition.size();
+  SplineProblem<> splineProblem(
+    Eigen::Vector2d(_currTime, _nextTime), _numCoefficients, numDofs);
+
+  assert(_currPosition.size() == numDofs);
+  assert(_nextPosition.size() == numDofs);
+  splineProblem.addConstantConstraint(0, 0, _currPosition);
+  splineProblem.addConstantConstraint(1, 0, _nextPosition);
+
+  if (_numCoefficients >= 4)
+  {
+    assert(_currVelocity.size() == numDofs);
+    assert(_nextVelocity.size() == numDofs);
+    splineProblem.addConstantConstraint(0, 1, _currVelocity);
+    splineProblem.addConstantConstraint(1, 1, _nextVelocity);
+  }
+
+  if (_numCoefficients >= 6)
+  {
+    assert(_currAcceleration.size() == numDofs);
+    assert(_nextAcceleration.size() == numDofs);
+    splineProblem.addConstantConstraint(0, 2, _currAcceleration);
+    splineProblem.addConstantConstraint(1, 2, _nextAcceleration);
+  }
+
+  const auto splineSegment = splineProblem.fit();
+  return splineSegment.getCoefficients()[0];
+}
+
+//=============================================================================
+void extractJointTrajectoryPoint(
+  const trajectory_msgs::JointTrajectory& _trajectory,
+  size_t _index, size_t _numDofs,
+  Eigen::VectorXd* _positions, bool _positionsRequired,
+  Eigen::VectorXd* _velocities, bool _velocitiesRequired,
+  Eigen::VectorXd* _accelerations, bool _accelerationsRequired)
+{
+  const auto& waypoint = _trajectory.points[_index];
+
+  try
+  {
+    checkVector("positions", waypoint.positions, _numDofs,
+      _positionsRequired, _positions);
+    checkVector("velocities", waypoint.velocities, _numDofs,
+      _velocitiesRequired, _velocities);
+    checkVector("accelerations", waypoint.accelerations, _numDofs,
+      _accelerationsRequired, _accelerations);
+  }
+  catch (const std::invalid_argument& e)
+  {
+    std::stringstream message;
+    message << "Waypoint " << _index << " is invalid: " << e.what();
+    throw std::invalid_argument(message.str());
+  }
+}
+
+// The rows of vector and matrix will be reordered using permutationMatrix
+void reorder(const Eigen::MatrixXd& permutationMatrix,
+  const Eigen::VectorXd& inVector,
+  const Eigen::MatrixXd& inMatrix,
+  Eigen::VectorXd* outVector,
+  Eigen::MatrixXd* outMatrix)
+{
+  *outVector = permutationMatrix*inVector;
+  *outMatrix = permutationMatrix*inMatrix;
+}
+
+// Convert the mapping to a permutation matrix for matrix operation
+void convertToPermutationMatrix(const std::map<size_t, size_t>& map,
+  Eigen::MatrixXd* permutationMatrix)
+{
+  *permutationMatrix = Eigen::MatrixXd::Zero(map.size(), map.size());
+  for(auto it = map.begin(); it != map.end(); ++it)
+  {
+    (*permutationMatrix)(it->second, it->first) = 1.0;
+  }
+}
+
+} // namespace
```

Comment:
I suspect that using a permutation matrix here will be _a lot_ slower than operating directly on the list of `std::pair<size_t, size_t>`s. I also think it is is more complicated to implement this way, since it involves an intermediate data structure (the permutation matrix) that would not be otherwise necessary.

Since this is an entirely an implementation detail, I am fine keeping it as it is for now. However, we may want to consider refactoring this in the future if performance becomes an issue.